### ISSUE TITLE
[HIGH] BERT-base (110M params) in V3 classifier — 2x slower than DistilBERT

### ISSUE THREAD

**Issue by Daksh7785** [riteshbonthalakoti/HELPDESK.AI#3119]
### File
`backend/services/classifier_v3.py`

### Location
```python
BertModel.from_pretrained("bert-base-uncased")
```

### Description

The V3 classification service currently loads the full **BERT-Base Uncased** model:

```python
BertModel.from_pretrained("bert-base-uncased")
```

BERT-Base contains approximately **110 million parameters**, making it substantially larger and more computationally expensive than lightweight alternatives such as **DistilBERT (~66M parameters)** or **MobileBERT**.

While BERT-Base provides strong language understanding capabilities, it is not optimized for low-latency inference in real-time API environments, particularly when running on CPU-only infrastructure.

The current implementation also does not appear to leverage inference optimizations such as:

- ONNX Runtime
- TorchScript compilation
- Dynamic quantization
- Model pruning
- Hardware acceleration

As a result, the classifier consumes excessive CPU, memory, and inference time, potentially degrading overall API performance.

### Current Behavior

The application initializes:

```python
model = BertModel.from_pretrained(
    "bert-base-uncased"
)
```

Inference executes using the full BERT architecture without optimization.

Approximate model sizes:

```text
BERT-Base      → ~110M parameters
DistilBERT     → ~66M parameters
MobileBERT     → ~25M parameters
```

### Performance Implications

Typical CPU inference characteristics:

```text
BERT-Base
→ ~2x slower than DistilBERT
→ ~2x higher memory usage
```

Estimated production impact:

```text
Higher latency
Higher RAM consumption
Reduced throughput
Increased infrastructure costs
```

Actual performance depends on:

- CPU specifications.
- Batch size.
- Sequence length.
- Concurrency levels.
- Hardware acceleration availability.

#### ❌ Increased API Latency

BERT-Base significantly increases inference time.

Potential consequences:

- Slow ticket analysis.
- Delayed API responses.
- Reduced user experience.

---

#### ❌ Excessive CPU Consumption

The larger model architecture requires more computation per request.

Consequences:

- Increased CPU utilization.
- Higher infrastructure costs.
- Reduced concurrent request capacity.

---

#### ❌ Higher Memory Usage

BERT-Base consumes substantially more RAM.

Potential effects:

- Increased container memory requirements.
- Reduced pod density.
- Greater risk of memory pressure.

---

#### ❌ Reduced Throughput

Because inference takes longer:

```text
Requests/second decrease
```

The application may struggle to support production traffic volumes.

---

#### ❌ Poor Suitability for CPU-Only Deployments

Deployments on:

- Small VMs
- Serverless platforms
- Containerized environments

may experience unacceptable inference latency.

### Root Cause

The classifier explicitly loads:

```python
BertModel.from_pretrained(
    "bert-base-uncased"
)
```

without applying any inference optimization techniques.

Current architecture:

```text
Full BERT Model
      +
No Quantization
      +
No ONNX Optimization
      +
CPU Inference
```

This combination results in unnecessary computational overhead.

### Suggested Fix

Replace BERT-Base with a lighter architecture optimized for inference workloads.

Recommended alternatives:

#### Preferred

```text
DistilBERT
```

Benefits:

- ~40% fewer parameters.
- Faster inference.
- Lower memory consumption.
- Comparable classification accuracy for many tasks.

Example:

```python
from transformers import DistilBertModel

model = DistilBertModel.from_pretrained(
    "distilbert-base-uncased"
)
```

---

#### Alternative

```text
MobileBERT
```

Benefits:

- Very low latency.
- Significantly reduced memory footprint.
- Suitable for constrained environments.

#### 1. Enable ONNX Runtime Optimization

Export the model:

```bash
python -m transformers.onnx
```

Run inference using:

```python
import onnxruntime as ort
```

Benefits:

- Faster inference.
- Lower CPU usage.
- Improved throughput.

---

#### 2. Apply Dynamic Quantization

Example:

```python
import torch

quantized_model = torch.quantization.quantize_dynamic(
    model,
    {torch.nn.Linear},
    dtype=torch.qint8
)
```

Expected improvements:

```text
Smaller model size
Lower memory usage
Faster CPU inference
```

---

#### 3. Use TorchScript

Compile the model:

```python
scripted_model = torch.jit.script(model)
```

Benefits:

- Reduced runtime overhead.
- Faster execution.

---

#### 4. Benchmark Candidate Models

Measure:

- Latency.
- Memory usage.
- Accuracy.
- Throughput.

before selecting the final production model.

Example metrics:

```text
P50 latency
P95 latency
Requests/sec
RAM usage
Classification accuracy
```

---

#### 5. Cache Models During Startup

Ensure the model is loaded once and reused across requests.

Avoid:

```python
model = load_model()
```

inside request handlers.

### Expected Outcome

- ✅ Inference latency decreases significantly.
- ✅ CPU utilization is reduced.
- ✅ Memory consumption decreases.
- ✅ API throughput improves.
- ✅ Infrastructure costs are lowered.
- ✅ The classification service becomes more suitable for production workloads.

### Priority

**High**

Although the current implementation may function correctly, using an unoptimized BERT-Base model introduces substantial performance and scalability overhead. Replacing it with a lightweight architecture and enabling inference optimizations is strongly recommended before production deployment.

**Comment by riteshbonthalakoti** [comment]
Hi! I have reopened this issue as requested.

⚠️ Please complete the GSSoC onboarding steps so that I can assign this issue to you:
1. 👤 **Follow the Project Admin**: https://github.com/ritesh-1918 (Mandatory)
2. ⭐ **Star this repository**: https://github.com/ritesh-1918/HELPDESK.AI
3. 🍴 **Fork this repository**: https://github.com/ritesh-1918/HELPDESK.AI/fork

💼 **Mandatory Testing Signup**:
Please sign up on our testing platform and choose **Ritesh PVT Limited** as your company. Once done, let me know here so I can assign the issue to you! 🚀

**Comment by riteshbonthalakoti** [comment]
Hi @Daksh7785! I have reopened this issue for you. Please complete the onboarding steps above so I can assign the issue to you. Let's build! 🚀

**Comment by riteshbonthalakoti** [comment]
Hi @Daksh7785! I have reopened this issue for you. Please complete the onboarding steps above so I can assign the issue to you. Let's build! 🚀
